File: 프로그래머스/입국심사/immigration.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
long long solution(int n, vector<int> times) {
	long long answer = 0;
	long long sum = 0;
	sort(times.begin(), times.end());
	long long min = 1;
	long long max = (long long)times.back() * n;
	long long mid = (max + min) / 2;
	answer = max;
	while (min <= max) {
		sum = 0;
		mid = (max + min) / 2;
		for (int i = 0; i < times.size(); i++) {
			sum += mid / times[i];
		}
		if (sum < n) min = mid + 1;
		else {
			if (mid <= answer)
				answer = mid;
			max = mid - 1;
		}
	}
	return answer;
}

File: 프로그래머스/입국심사/immigration.java
class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        long low = 1;
        long high = (long)n* 1000000000;
        while(low<high){
            long mid = (low+high)/2;
            long totalcount=0;
            for(int i=0; i<times.length;i++){
                totalcount += mid/times[i];
                if(totalcount>n) break;
            }
            if(totalcount>=n)
                high = mid;
            else
                low = mid+1;
        }
        answer = low;
        return answer;
    }
}